### PR TITLE
CP-21576: Add Config command for Listing Kubernetes Services

### PR DESCRIPTION
**Pull request by bdrennz**
This PR adds a new command, `list-services`, to the `config` suite. The command finds all the active services within the a given Cluster. Users have the option to specify a `--kubeconfig`, or by the configuration found in the current directory.

Testing:
``` bash
(system) ➜  cloudzero-agent-validator git:(cp-21576) ./bin/cloudzero-agent-validator config list-services --kubeconfig /tmp/bdren.kubeconfig                                                                                                        <aws:cz-research.Engineering> <region:us-east-1>(⎈|arn:aws:eks:us-east-1:975482786146:cluster/bdren:default)
Services in all namespaces:
 - cz-prom-agent-kube-state-metrics (Namespace: default)
 - cz-prom-agent-prometheus-node-exporter (Namespace: default)
 - kubernetes (Namespace: default)
 - cz-prom-agent-kube-prometh-kubelet (Namespace: kube-system)
 - kube-dns (Namespace: kube-system)
 - metrics-server (Namespace: kube-system)
 - my-otel-demo-adservice (Namespace: otel)
 - my-otel-demo-cartservice (Namespace: otel)
 - my-otel-demo-checkoutservice (Namespace: otel)
 - my-otel-demo-currencyservice (Namespace: otel)
 - my-otel-demo-emailservice (Namespace: otel)
 - my-otel-demo-featureflagservice (Namespace: otel)
 - my-otel-demo-ffspostgres (Namespace: otel)
 - my-otel-demo-frontend (Namespace: otel)
 - my-otel-demo-frontendproxy (Namespace: otel)
 - my-otel-demo-grafana (Namespace: otel)
 - my-otel-demo-jaeger-agent (Namespace: otel)
 - my-otel-demo-jaeger-collector (Namespace: otel)
 - my-otel-demo-jaeger-query (Namespace: otel)
 - my-otel-demo-kafka (Namespace: otel)
 - my-otel-demo-loadgenerator (Namespace: otel)
 - my-otel-demo-otelcol (Namespace: otel)
 - my-otel-demo-paymentservice (Namespace: otel)
 - my-otel-demo-productcatalogservice (Namespace: otel)
 - my-otel-demo-prometheus-server (Namespace: otel)
 - my-otel-demo-quoteservice (Namespace: otel)
 - my-otel-demo-recommendationservice (Namespace: otel)
 - my-otel-demo-redis (Namespace: otel)
 - my-otel-demo-shippingservice (Namespace: otel)
 - otel-demo-opensearch (Namespace: otel)
 - otel-demo-opensearch-headless (Namespace: otel)

```

### CODE REVIEW

File: pkg/config/prometheus.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ func (s *Prometheus) Validate() error {
 	}
 
 	if s.PrometheusNodeExporterServiceEndpoint == "" {
-		return fmt.Errorf(ErrNoPrometheusNodeExporterServiceEndpointMsg)
+		return errors.New(ErrNoPrometheusNodeExporterServiceEndpointMsg)
```
Comment (bdrennz) [inline pkg/config/prometheus.go:26]: Lint fix

File: pkg/cmd/config/command.go [hunk 2]
```diff
@@ -100,13 +126,38 @@ func Generate(values map[string]interface{}, outputFile string) error { //nolint
 	return t.Execute(out, values)
 }
 
+// ListServices lists all Kubernetes services in all namespaces
+func ListServices(clientset kubernetes.Interface) error {
```
Comment (josephbarnett) [inline pkg/cmd/config/command.go:130]: We might want to move this to it's own `pkg` (aka library in every other language).

I might recommend https://github.com/Cloudzero/cloudzero-agent-validator/tree/develop/pkg/diagnostic/k8s

or even a new directory where we put k8s API things such as https://github.com/Cloudzero/cloudzero-agent-validator/tree/develop/pkg/k8s.

File: pkg/cmd/config/command.go [hunk 3]
```diff
@@ -73,6 +78,27 @@ func NewCommand() *cli.Command {
...
+					kubeconfigPath := c.String("kubeconfig")
+					config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+					if err != nil {
+						return errors.Wrap(err, "building kubeconfig")
+					}
+
+					clientset, err := kubernetes.NewForConfig(config)
+					if err != nil {
+						return errors.Wrap(err, "creating clientset")
+					}
```
Comment (josephbarnett) [inline pkg/cmd/config/command.go:97]: We might make this into a function also. 

I think there is something [similar here](https://github.com/Cloudzero/cloudzero-agent-validator/blob/develop/pkg/diagnostic/k8s/check.go#L82-L89)

**Review by josephbarnett — APPROVED** [review]
a few nits - but generally pretty cool! 🙌

File: pkg/cmd/config/command.go (hunk 2 above)
Comment (bdrennz) [reply pkg/cmd/config/command.go:130]: Agreed, much tighter having it in it's own package. I'll make the update.

File: pkg/cmd/config/command.go (hunk 3 above)
Comment (bdrennz) [reply pkg/cmd/config/command.go]: Awesome thanks, just updated.

File: pkg/diagnostic/k8s/check.go [hunk 4]
```diff
@@ -22,20 +22,20 @@ import (
 
 const DiagnosticK8sVersion = config.DiagnosticK8sVersion
 
-type checker struct {
+type Checker struct {
```
Comment (bdrennz) [inline pkg/diagnostic/k8s/check.go:25]: Allowing `Checker`  to be exported

File: pkg/diagnostic/k8s/check.go [hunk 5]
```diff
@@ -76,10 +76,10 @@ func (c *checker) getK8sVersion(_ *http.Client) ([]byte, error) {
 	return []byte(fmt.Sprintf("%s.%s", information.Major, information.Minor)), nil
 }
 
-// getConfig returns a k8s config based on the environment
+// GetConfig returns a k8s config based on the environment
```
Comment (bdrennz) [inline pkg/diagnostic/k8s/check.go:79]: Allowing GetConfig to be exported

File: pkg/diagnostic/k8s/check.go (hunk 4 above)
Comment (josephbarnett) [reply pkg/diagnostic/k8s/check.go:25]: Actually you don't want to break this encapsulation. Notice the "NewProvider" passed back and interface with well known functions. 

I would recommend making a new interface is needed; then cast the returned pointer to that interface.


That said, I would also recommend moving `GetConfig` to a more generic method in the k8s package you just created.

**Pull request merged**